src/hyp_map.rs

```diff
@@ -70,18 +72,122 @@ impl SharedRegion {
             let vaddr = PageAddr::new(RawAddr::supervisor_virt(r.base().bits())).unwrap();
             let page_count = PageSize::num_4k_pages(r.size()) as usize;
             let pte_fields = PteFieldBits::leaf_with_perms(pte_perms);
+            let variable_mapped = matches!(
+                r.region_type(),
+                HwMemRegionType::Reserved(HwReservedMemType::HypervisorHeap)
+                    | HwMemRegionType::Reserved(HwReservedMemType::HypervisorPerCpu)
+            );
             Some(Self {
                 vaddr,
                 paddr,
                 page_count,
                 pte_fields,
+                variable_mapped,
+                page_size: None,
             })
         } else {
             None
         }
     }
 
-    // Map this region into a page table.
+    /// partition a HypMapRegion to use minimal pages
+    pub fn partition(&self) -> Option<SharedRegionsVec> {
```

Comment:
This logic is pretty confusing to follow and there are a couple edge cases that I think aren't quite handled correctly (but need to read again more carefully).

First, I'd recommend doing this iteratively in `map()`; there's really no need to recursively create an intermediate vector of regions.

Next, I think a more concise (and easier to read) way to structure this would be something like:

```
fn select_page_size(va: u64, pa: u64, len: u64) -> PageSize {
    for size in [Size512G, Size1G, Size2m] {
        if size.is_aligned(va) && size.is_aligned(pa) && len >= (size as u64) {
            return size;
        }
    }

    return Size4k;
}

fn map(...) {
    let mut offset = 0;
    while offset != size {
        let remaining = size - offset;
        let va = self.vaddr() + offset;
        let pa = self.paddr() + offset;
        let page_size = select_page_size(va, pa,, remaining);
        let num_pages = remaining / (page_size as u64);
        let mapper = map_range(va, page_size, num_pages, get_pte_page);
        for (v, p) in va.iter_from_with_size(page_size).zip(...) {
            mapper.map_addr(...);
        }
        offset += page_size as u64 * num_pages;
    }
}
```